### ISSUE TITLE
[share_plus]  share() result in empty message in Facebook News Feed

### ISSUE THREAD

**Issue by Ciock** [fluttercommunity/plus_plugins#413]
## System info

Issue occurs on: both
Plugin name: share_plus
Plugin version: 2.1.4
Flutter doctor output: 

```
[√] Flutter (Channel beta, 2.4.0-4.2.pre, on Microsoft Windows [Versione 10.0.19043.1165], locale it-IT)
[√] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[√] Chrome - develop for the web
[√] Android Studio (version 4.1.0)
[√] IntelliJ IDEA Ultimate Edition (version 2020.1)
[√] VS Code (version 1.59.0)
[√] Connected device (3 available)

• No issues found!
```

## Steps to Reproduce

1. `Share.share('text');`
2. Select Facebook News Feed
3. Then the Facebook post textfield is empty with the placeholder _"What's on your mind?"_

It is quite a serious problem because I have been contacted by the Facebook Team regarding this feature which does not respect their policies:

> 
> Developer Policy 1.2: Build an app that is stable and easily navigable.Some common violations of this policy include:
>   - Broken Facebook integration (e.g. broken share, like or comment functionality)
>   - Broken user experience in the app (e.g. app has broken links or user interface failures)
>

**Comment by Boehrsi** [comment]
I did some research regarding this topic (if someone knows other / more information please correct me) and it seems like it's not working with this plugin (or with the whole native Android sharing concept in general) as Facebook decided to not implement / block the normal sharing flow. Keeping this in mind it's actually hilarious that Facebook is blaming you that stuff is not working.

My findings:
- Facebook doesn't support [share via Intent](https://developer.android.com/training/sharing/send) which is the Android way of sharing content
- Facebook has it's own SDK with sharing functionality https://developers.facebook.com/docs/sharing/ / https://github.com/facebook/facebook-android-sdk 
- There are other plugins handling the Facebook specific part (not checked for quality just listing some) and all of them require e.g. a Facebook SDK specific key
  - https://pub.dev/packages/flutter_share_me
  - https://pub.dev/packages/flutter_social_content_share
  - https://pub.dev/packages/social_share_plugin

Even though Facebook is a big player as company and in the internet, this is really really bad!

The SDK contains a lot of stuff and potentially analytics and what so ever. 
@mhadaily I would highly vote (even though I potentially have no vote here 😄) against including such a potentially big dependency and app specific code just because a company decides it's big enough to not support the normal default workflows. 
@Ciock I think the best approach for you would be evaluating one or another of the mentioned plugins if those support your use cases.

**Comment by jasonJamEther** [comment]
I am getting the same result on ios when sharing to facebook messenger. `Share.shareFiles([file.path], text: 'share text goes here');`
results in the image appearing as expected in messenger but the message text is empty.

**Comment by Boehrsi** [comment]
You are welcome @Ciock I'm happy I could provide some information and I'm sorry that the other libraries aren't helpful.

@mhadaily Could it be a good idea to document this topic? I could provide a PR to avoid more confusion in this area.

**Comment by vinnytwice** [comment]
HI I'm getting to code the sharing part for my app but I don't yet really master the whole process..
 I'm a bit confused about how to handle the sharing problem with Facebook as experienced by @Ciock. If I got it right this problem is only on Android as with `share_plus` package the sharing would be carried out via Intent but Facebook doesn't accept them and any content share will cause problems with Facebook ( eg brake the dev policy as mentioned by @Ciock) when Facebook is the app receiving the Intent. 
So if  Facebook is gonna give problems to everybody wouldn't be possible to either bridge native Facebook SDK or exclude Facebook from sharing capabilities in this package with `Intent.EXTRA_EXCLUDE_COMPONENTS` or other ways? Otherwise the whole sharing on Flutter gets impossible to implement.. well at least in a native looking manner.
I saw that the other mentioned packages have their defined methods to be called for specific apps..so you can't use the Android Sharesheet and need to build a custom UI for sharing which is rather ugly and discouraged practice from Android docs..

**Comment by Boehrsi** [comment]
@vinnytwice hi, I'm trying to summarize the situation, but I'm myself haven't had a look into Facebooks policies and in which situations they contact you and claim something is "wrong" (potentially if you integrate something with their login / some Facebook specific functionality / their SDKs and so on).

- share_plus for Android (I'm not an iOS developer and can't say something about that area) uses the normal system intent way to share data with the system and so other apps. Therefore an intent is send to the system, which contains some fields like `EXTRA_TEXT`. The receiving app can handle those. 
- Every app can decide how to handle the data (some best practices are mentioned here https://developer.android.com/training/sharing/send).
- Facebook denies to handle some of the cases and they offer their own SDK instead. With this SDK you then can share data. The SDK requires some API key and therefore a registration or something similar with Facebook.
- In regards to "the whole sharing in Flutter gets impossible to implement". I understand what you mean, but not everybody integrates Facebook specific code or handling, so the scope would be "only" (and I'm aware Facebook is a big player and there are multiple use cases targeting Facebook in regards to sharing) Facebook related apps.
- Building an own UI and logic for sharing is pretty bad, you are totally right and that should be avoided for sure. I'm not sure how Facebook defines in their guides and policies how to integrate the Facebook SDK with a normal native Android app. But the Flutter approach would be to follow this native approach and then trigger the flow from within the Flutter world via method channels. In general exactly what the share_plus plugin is already doing to trigger the system intent flow.
- The Intent.EXTRA_EXCLUDE_COMPONENTS idea seems pretty neat, but I'm unsure if "disable sharing with Facebook at all" would also be considered by Facebook as "broken share functionality". But this is just conjecture and potentially it would be a good idea if someone would contact Facebook in regards to that.
- The problem with the Facebook SDK is that it’s a potentially huge overhead, which forces user to register API keys, use a bigger library with a proprietary SDK and so on. That’s the main reason why it’s not in this general purpose plugin (at least that’s my understanding). It could be an idea to write something like an extension for share_plus, which adds this functionality on top for users who are especially interested in this sharing area.

**Comment by vinnytwice** [comment]
@Boehrsi Hi and thanks for answering so quickly. With "the whole sharing in Flutter gets impossible to implement" I actually meant that if you don't exclude Facebook from the package somehow, then it will be available to system to be targeted for sharing if installed on user's device, hence affecting the entire sharing content implementation, that's why I thought of the Intent.EXTRA_EXCLUDE_COMPONENTS. From a package designer prospective I think this should be something to take big consideration in otherwise this would be jus an "almost working" package. I looked at how other packages handle this and the extension route is very feasible and dough unfortunately I'm not the man for the job at the moment, as a side project I'll be looking into it. A great package like share_plus shouldn't be killed by such a bully company.. I'll surely find out if Intent.EXTRA_EXCLUDE_COMPONENTS break some Facebook dev policy as it my be a temporary solution while an extension is made. Thank you again. Cheers.

**Comment by vbuberen** [comment]
Would like to add here, that it is not a `almost working` package, but Facebook breaking the rules and guidelines of Android. In case Google would be more strict about it we wouldn't have such problem.

> So if Facebook is gonna give problems to everybody wouldn't be possible to either bridge native Facebook SDK

From my point of view it shouldn't even be considered, because this is not just adding already mentioned burden of doing some setup by users it is also a huge risk to stability of all apps which will use `share_plus` because Facebook is known for their bad quality https://9to5mac.com/2020/07/10/app-crash-facebook-sdk/  . Not to mention that it would be too much to bend over to make artificially complicated products from a very unethical company work well with `share_plus` plugin, which works great with all the apps, which respect platform guidelines they are working on.

**Comment by Boehrsi** [comment]
@vinnytwice thanks for looking into Intent.EXTRA_EXCLUDE_COMPONENTS and the policies, this could be something share_plus could add (as configurable option) to help users out here. 
The independent extension with Facebook support is something I would sadly also be not the right person for to implement, so also thanks in that area that you will have a look. Even as a side project your insights could be pretty helpful and possibly someone else gets motivated and helps out.

@vbuberen thanks for the additional information and I totally agree here, share_plus should never force developers per default to add proprietary and potentially even bad quality code / SDKs.

**Comment by jamesncl** [comment]
I would agree that it's the right decision not to modify this package to support Facebook's "sharing" requirements. As developers we should be supporting openness and interoperability wherever possible, something that this would directly work against.

**Comment by cedvdb** [comment]
I'm all for openness and all, but there is the sad reality that they are one of the biggest player and not having that functionality is a big loss for some, or just not meeting requirements to others. If everyone took the stand to not use their share sdk I'm sure they'd bend but I don't think that's going to happen anytime soon.

**Comment by cedvdb** [comment]
I don't understand the stance here.

 - So the package won't correctly work with Facebook, messenger, instagram and whatsapp ?
 - There is currently no way of excluding those platforms with Intent.EXTRA_EXCLUDE_COMPONENTS

So should the package:
  - not be used in production ?
  - or should not be used if you intend to not break Facebook TOS ?

**Comment by aytunch** [comment]
@Boehrsi does this limitation Facebook is bringing have anything to do with this issue?
I agree that what facebook does here is not fair play.
However, if adding a facebook app id and secret key solves these issues, I don't see why we don't integrate it in this plugin?
Am I missing something?
https://developers.facebook.com/docs/instagram/sharing-to-stories
https://developers.facebook.com/blog/post/2022/10/10/introducing-important-update-to-Instagram-sharing-to-stories/

**Comment by miquelbeltran** [comment]
Locking this conversation.

Adding Facebook SDK is out of question for this plugin and will never be considered.

Please use a plugin that uses the Facebook SDK instead.
